### PR TITLE
feat: permit2 allowance flow for zrx swaps

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Breaks token allowance approval UI components into reusable ones in prep for permit2 UI flow.

Notable non-permit2 changes:

- Adds default "exact" allowance state per conversation with product (previously this was only the case for LiFi trades).
- Copy for allowance reset is now "allowance reset", not "approval reset"
- Copy for allowance approval is not "allowance approval", not "approval"
- Code renaming to correspond to the above
- Combine allowance reset and allowance approval steps into a single allowance step component
- Migrates 0x swapper to use api proxy
- UI will persistently display all approval related TX hashes (e.g both the allowance reset and allowance approval), with better naming

Permit2 changes:

- Fully implements Permit2 flow, which consists of an initial unlimited allowance approval (toggle will be hidden) followed by a signature
- If the initial approval was already processed for the sell token, only a signature is required
- USDT allowance reset never appears for Permit2 because only unlimited allowance is presented when initially approving. If a user gets into a state they need to reset their allowance, they will have to use https://app.unrekt.net/
- A feature toggle to disable the Permit2 version of the 0x swapper has been included so we can fall back to the v1 API if we have any issues. Due to the surface area of the change, no feature toggle exists for the UI and state changes so extra care is required to ensure we roll this out without issues. 

TODO:

- [x] Beard oil
- [x] ETH -> ERC20 broken
- [x] insufficient <buy asset> for fees
- [x] triple check zrx still works with flag off
- [x] triple check zrx v2 api proxy endpoints are deployed to prod

## Issue (if applicable)

progresses #7648

## Risk

> High Risk PRs Require 2 approvals

This is a high risk change because it heavily modifies the state management of trades and completely rewrites the UI associated with allowance approvals.

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

All trades, especially those with token allowance approvals. Due to the surface area of the change, no feature toggle exists for the UI and state changes so extra care is required to ensure we roll this out without issues. 

### General

- [ ] trades on all chains

### non 0x trades

- [ ] without allowance approvals
- [ ] trades with allowance approvals
- [ ] trades with USDT allowance reset and approval
- [ ] multi-hop trades

### 0x trades

- with `REACT_APP_FEATURE_ZRX_PERMIT2` feature flag enabled
  - [ ] without allowance approvals
  - [ ] trades with allowance approvals
  - [ ] trades with USDT allowance reset and approval

- with `REACT_APP_FEATURE_ZRX_PERMIT2` feature flag disabled
  - [ ] trades with initial Permit2 allowance approval + sign
  - [ ] trades with Permit2 sign only (i.e allowance previously approved)

### Operations

There is a feature flag in this PR but its used only as a fallback if we need to disable permit2 trades for any reason. The changes made in this PR still affect the app majorly regardless of whether the flag is enabled.

### non-0x trades
  - without allowance approvals
    https://jam.dev/c/27122198-3e0e-4669-84a9-37bdcb76fc52
  - trades with allowance approvals
    https://jam.dev/c/0aadbb0a-681a-400d-86bb-80bc6a9e36e5
  - trades with USDT allowance reset and approval
    https://jam.dev/c/671e6347-c888-4b11-95ab-0275728a1729

### 0x trades
- trades with initial Permit2 allowance approval + sign
  https://jam.dev/c/7c1535c7-1966-4ed0-807d-40ef332b46f6
- trades with Permit2 sign only (i.e allowance previously approved)
  https://jam.dev/c/d6d8af5d-0e6d-486d-90bf-fcb4e3976bf5

### CODE REVIEW

File: src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/helpers.ts [hunk 1]
```diff
@@ -28,3 +30,10 @@ export const approveTrade = async ({
 
   return txHash
 }
+
+export const isPermit2Hop = (tradeQuoteStep: TradeQuoteStep | undefined) => {
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/helpers.ts:34]: Can we TODO this to generalize it vs. SwapperName heuristics? 🙏🏽

Comment (gomesalexandre) [reply src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/helpers.ts:34]: Or perhaps SwapperName *is* the right heuristics as we add Permit2 to other swappers?

**Review by gomesalexandre — COMMENTED** [review]
First very very high-level pass to get familiar with this - untested at runtime just yet and allowance hooks still not reviewed, but looks solid from this first read!
Some comments re: generalization and naming mostly, though nothing obviously stands out as wrong here.

**Review by gomesalexandre — COMMENTED** [review]
First runtime pass with permit2 flag enabled

  - RUNE -> AVAX (THOR) - note Tx history is sad but irrelevant to this PR, got a Zerion notification that fundus were received
  https://jam.dev/c/96cc06bb-74fd-4257-9bed-e694e20d4aa2
 - XDAI -> FOX (Li.Fi)
 https://jam.dev/c/af077d67-8da7-4513-8206-f1970109de02
  - FOX -> XDAI (CoW, no allowance granted)
  https://jam.dev/c/34b13a10-c595-4f45-905c-9701b73571a7
  - FOX -> XDAI (CoW, allowance granted)
  https://jam.dev/c/4ff9213d-c873-4da6-beb2-d046be67fd0b
  - Shitcoin -> ETH bridge (Li.Fi, no allowance granted)
  https://jam.dev/c/5b65b4af-3d15-4bb8-8320-cb1645b0774e
  - FOX -> ETH (0x) quote 🚫 
  Getting this error response back from ZRX
  image
  - USDT.ARB -> ETH.ARB (0x) quote 🚫 
  Same here:
  image

  
Also tested with the flag off:

- USDT.AVA -> AVA.AVA (0x, no approval granted)
https://jam.dev/c/8089bdbc-40d3-4df5-96b5-ffdd763162e3
- USDT.AVA -> AVA.AVA (0x, approval granted)
https://jam.dev/c/700cb7f9-bb00-4ecb-b3ca-bc00ac057312
- ETH -> USDT (Li.Fi)
https://jam.dev/c/3b55d215-c082-4574-9e93-297caf73d6d0

Unable to test more USDT flow re: unlimited, 0, and approval reset flows for the time being because of low fundus

**Review by gomesalexandre — COMMENTED** [review]
Tested locally and LGTM for all tested flows, minus one spotted issue re: permit2 contract approval being prompted (with the flag on) when there is enough allowance, and there shouldn't be any new allowance Tx, hence no permit2 contract approval nor actual allowance Tx/permit2 EIP712 message signed and broadcasted.

See testing in https://gist.github.com/gomesalexandre/6c0b209f940b64a09d726c53ad7f7527

File: src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/helpers.ts (hunk 1 above)
Comment (woodenfurniture) [reply src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/helpers.ts:5]: Yeah until we know what our next expansion of this is, we should wait and see

File: packages/contracts/src/viemClient.ts [hunk 2]
```diff
@@ -117,6 +117,18 @@ export const viemClientByChainId: Record<ChainId, PublicClient> = {
   [KnownChainIds.BaseMainnet]: viemBaseClient,
 }
 
+export const viemNetworkIdByChainId: Record<ChainId, number> = {
```
[the remark this answers is not in the document]
Comment (woodenfurniture) [reply packages/contracts/src/viemClient.ts:120]: I've opted for the viem one because it is a direct inverse of the one below this, `viemClientByNetworkId`, rather than having to look for a home for a caip version of the same thing

File: packages/swapper/src/types.ts [hunk 3]
```diff
@@ -292,10 +299,12 @@ export type CommonGetUnsignedTransactionArgs = {
 }
 
 export type GetUnsignedEvmTransactionArgs = CommonGetUnsignedTransactionArgs &
-  EvmAccountMetadata & { supportsEIP1559: boolean } & Omit<
-    EvmSwapperDeps,
-    'fetchIsSmartContractAddressQuery'
-  >
+  EvmAccountMetadata &
+  Omit<EvmSwapperDeps, 'fetchIsSmartContractAddressQuery'> & {
+    permit2Signature: string | undefined
```
[the remark this answers is not in the document]
Comment (woodenfurniture) [reply packages/swapper/src/types.ts:287]: We dont know yet, but likely yes as it uses the uniswap Permit2 contract. This is likely another "wait as see"

File: src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useTradeExecution.tsx [hunk 4]
```diff
@@ -259,6 +271,7 @@ export const useTradeExecution = (
             slippageTolerancePercentageDecimal,
             from,
             supportsEIP1559,
+            permit2Signature: permit2.permit2Signature,
```
[the remark this answers is not in the document]
Comment (woodenfurniture) [reply src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useTradeExecution.tsx:274]: Likely yes again, this is not a 0x specific implementation as it's using the generic uniswap Permit2 contract to handle all of this

File: packages/swapper/src/swappers/ZrxSwapper/utils/helpers/helpers.ts [hunk 5]
```diff
@@ -1,37 +1,99 @@
...
-      return 'https://0x.shapeshift.com/base/'
+      return `${zrxBaseUrl}base/`
     default:
       assertUnreachable(chainId)
   }
 }
 
 // converts an asset to zrx token (symbol or contract address)
-export const assetToToken = (asset: Asset): string => {
+export const assetToZrxToken = (asset: Asset): string => {
```
[the remark this answers is not in the document]
Comment (woodenfurniture) [reply packages/swapper/src/swappers/ZrxSwapper/utils/helpers/helpers.ts:34]: yeah nice

File: src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useSignPermit2.tsx [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import { useWallet } from 'hooks/useWallet/useWallet'
+import { assertGetEvmChainAdapter } from 'lib/utils/evm'
+import { selectPortfolioAccountMetadataByAccountId } from 'state/slices/selectors'
+import { selectHopSellAccountId } from 'state/slices/tradeQuoteSlice/selectors'
+import { tradeQuoteSlice } from 'state/slices/tradeQuoteSlice/tradeQuoteSlice'
+import { useAppDispatch, useAppSelector } from 'state/store'
+
+// handles allowance approval tx execution, fees, and state orchestration
+export const useSignPermit2 = (
+  tradeQuoteStep: TradeQuoteStep,
```
[the remark this answers is not in the document]
Comment (woodenfurniture) [reply src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useSignPermit2.tsx:15]: Prefer we wait for a need to decouple this, thinking in yagni terms. Since this hook is reasonably at the "leaf" of the swapper context, this is a two-way door and can be broken out on demand if there's ever a need to 🙏

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Code is looking solid, permit flow is solid.



I noticed one blocking regression whilst testing - on this branch we only seem able to get 0x quotes for the Ethereum chain. See screenshots below (left is this branch, right is production):

Screenshot 2024-10-18 at 16 03 36
Screenshot 2024-10-18 at 16 11 58

Nitpick: the tooltip being after a `:` is odd:

Screenshot 2024-10-18 at 16 09 46

File: packages/contracts/src/viemClient.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply packages/contracts/src/viemClient.ts:120]: Absolutely fine by me, ty for rationale!

File: packages/swapper/src/swappers/ZrxSwapper/utils/fetchFromZrx.ts [hunk 7]
```diff
@@ -71,3 +83,51 @@ export const fetchFromZrx = async <T extends 'price' | 'quote'>({
 
   return Ok(maybeZrxPriceResponse.unwrap().data)
 }
+
+export const fetchFromZrxPermit2 = async <T extends 'price' | 'quote'>({
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ZrxSwapper/utils/fetchFromZrx.ts:87]: We don't ever consume this with `price`.
Two qs: 
1. Is refetching an actual quote in `getUnsignedEvmTransaction` something we want, and can even do given the ramifications of how permit2 flow works?
2. Do we want to only get a rate quote in `getZrxTradeQuote`, or same thing, that won't work given how permit2 works?

File: packages/swapper/src/swappers/ZrxSwapper/getZrxTradeQuote/getZrxTradeQuote.ts [hunk 8]
```diff
@@ -154,3 +169,196 @@ export async function getZrxTradeQuote(
...
+    return Err(
+      makeSwapErrorRight({
+        message: `cross-chain not supported - both assets must be on chainId ${sellAsset.chainId}`,
+        code: TradeQuoteError.CrossChainNotSupported,
+        details: { buyAsset, sellAsset },
+      }),
+    )
+  }
+
+  const maybeZrxPriceResponse = await fetchFromZrxPermit2({
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ZrxSwapper/getZrxTradeQuote/getZrxTradeQuote.ts:228]:
```suggestion
  const maybeZrxQuoteResponse = await fetchFromZrxPermit2({
```

See https://github.com/shapeshift/web/pull/7758#discussion_r1806278116 for reference

File: src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useSignPermit2.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useSignPermit2.tsx:15]: yagni and on-demand sounds good to me, we may not need it indeed!

**Review by gomesalexandre — APPROVED** [review]
Retested locally as a last paranoia pass after last changes and confirmed we're looking good at runtime:

https://gist.github.com/gomesalexandre/ab10310c3aff40bd37c00fb15c3ed4da

One q: re price or quote logic for permit2, but otherwise looks sane to land :fridaydog:

File: src/components/MultiHopTrade/components/MultiHopTradeConfirm/components/Hop.tsx [hunk 9]
```diff
@@ -155,12 +155,14 @@ export const Hop = ({
...
+      // fallthrough
+      case HopExecutionState.AwaitingAllowanceApproval:
+      // fallthrough
+      case HopExecutionState.AwaitingPermit2:
         return hopIndex === 0 ? 1 : 0
-      case HopExecutionState.AwaitingApproval:
-        return hopIndex === 0 ? 2 : 1
       case HopExecutionState.AwaitingSwap:
-        return hopIndex === 0 ? 3 : 2
+        return hopIndex === 0 ? 2 : 1
```
[the remark this answers is not in the document]
Comment (woodenfurniture) [reply src/components/MultiHopTrade/components/MultiHopTradeConfirm/components/Hop.tsx:165]: No, this is required as there are more steps in the stepper state machine

File: packages/swapper/src/swappers/ZrxSwapper/utils/fetchFromZrx.ts (hunk 7 above)
Comment (woodenfurniture) [reply packages/swapper/src/swappers/ZrxSwapper/utils/fetchFromZrx.ts:87]: Gm:
1. Refetching is only done in `getUnsignedEvmTransaction` when `transactionMetadata` is absent from the quote, indicating the quote is from the V1 api. In this case, we'd need to fetch a fresh quote from the V1 api to obtain the transaction metadata. This will be removed once V2 api has been stable in prod for ~2 weeks.
2. When V1 api is removed, we can refactor to simplify everything. V2 flow uses only a single fetch of a `quote`, as opposed to the v1 flow which needs a `price` and then a `quote`.

TL;DR good observation and we'll clean up when we are confident we dont want to toggle back to the V2 api.

File: packages/swapper/src/swappers/ZrxSwapper/getZrxTradeQuote/getZrxTradeQuote.ts (hunk 8 above)
Comment (woodenfurniture) [reply packages/swapper/src/swappers/ZrxSwapper/getZrxTradeQuote/getZrxTradeQuote.ts:228]: Per reply above, we can come back and simplify everything when removing the V1 api.

**Review by 0xApotheosis — APPROVED** [review]
My nitpick feedback was addressed, and I'm putting the quote issues I had down to dev endpoint shenanigans.

gm
